File: wear/src/main/java/com/atakmap/android/wickr/plugin/multisensortracking/HeartRateListener.java
/*
 * Copyright 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.atakmap.android.wickr.plugin.multisensortracking;

import android.util.Log;

import androidx.annotation.NonNull;

import com.atakmap.android.wickr.plugin.R;
import com.samsung.android.service.health.tracking.HealthTracker;
import com.samsung.android.service.health.tracking.data.DataPoint;
import com.samsung.android.service.health.tracking.data.ValueKey;

import java.util.List;

public class HeartRateListener extends BaseListener {
    private final static String APP_TAG = "HeartRateListener";

    HeartRateListener() {
        final HealthTracker.TrackerEventListener trackerEventListener = new HealthTracker.TrackerEventListener() {
            @Override
            public void onDataReceived(@NonNull List<DataPoint> list) {
                for (DataPoint dataPoint : list) {
                    readValuesFromDataPoint(dataPoint);
                }
            }

            @Override
            public void onFlushCompleted() {
                Log.i(APP_TAG, " onFlushCompleted called");
            }

            @Override
            public void onError(HealthTracker.TrackerError trackerError) {
                Log.e(APP_TAG, " onError called: " + trackerError);
                setHandlerRunning(false);
                if (trackerError == HealthTracker.TrackerError.PERMISSION_ERROR) {
                    TrackerDataNotifier.getInstance().notifyError(R.string.NoPermission);
                }
                if (trackerError == HealthTracker.TrackerError.SDK_POLICY_ERROR) {
                    TrackerDataNotifier.getInstance().notifyError(R.string.SdkPolicyError);
                }
            }
        };
        setTrackerEventListener(trackerEventListener);
    }

    public void readValuesFromDataPoint(DataPoint dataPoint) {
        final HeartRateData hrData = new HeartRateData();
        final List<Integer> hrIbiList = dataPoint.getValue(ValueKey.HeartRateSet.IBI_LIST);
        final List<Integer> hrIbiStatus = dataPoint.getValue(ValueKey.HeartRateSet.IBI_STATUS_LIST);

        hrData.status = dataPoint.getValue(ValueKey.HeartRateSet.HEART_RATE_STATUS);
        hrData.hr = dataPoint.getValue(ValueKey.HeartRateSet.HEART_RATE);
        if (hrIbiList != null && hrIbiList.size() != 0) {
            hrData.ibi = hrIbiList.get(hrIbiList.size() - 1); // Inter-Beat Interval (ms)
        }
        if (hrIbiStatus != null && hrIbiStatus.size() != 0) {
            hrData.qIbi =  hrIbiStatus.size() - 1; // 1: bad, 0: good
        }
        TrackerDataNotifier.getInstance().notifyHeartRateTrackerObservers(hrData);
        Log.d(APP_TAG, dataPoint.toString());
    }

}

File: wear/src/main/java/com/atakmap/android/wickr/plugin/tracking/HeartRateListener.kt
package com.atakmap.android.wickr.plugin.tracking

import com.atakmap.android.wickr.plugin.R
import com.samsung.android.service.health.tracking.HealthTracker.TrackerError
import com.samsung.android.service.health.tracking.HealthTracker.TrackerEventListener
import com.samsung.android.service.health.tracking.data.DataPoint
import com.samsung.android.service.health.tracking.data.ValueKey
import org.koin.core.component.KoinComponent
import org.koin.core.component.get

class HeartRateListener internal constructor() : BaseListener(), KoinComponent {

    private val trackerDataNotifier: TrackerDataNotifier = get()

    init {
        val trackerEventListener: TrackerEventListener = object : TrackerEventListener {
            override fun onDataReceived(list: List<DataPoint>) {
                for (dataPoint in list) {
                    readValuesFromDataPoint(dataPoint)
                }
            }

            override fun onFlushCompleted() {

            }

            override fun onError(trackerError: TrackerError) {
                setHandlerRunning(false)
                if (trackerError == TrackerError.PERMISSION_ERROR) {
                    trackerDataNotifier.notifyError(R.string.NoPermission)
                }
                if (trackerError == TrackerError.SDK_POLICY_ERROR) {
                    trackerDataNotifier.notifyError(R.string.SdkPolicyError)
                }
            }
        }

        setTrackerEventListener(trackerEventListener)
    }

    fun readValuesFromDataPoint(dataPoint: DataPoint) {
        val hrData = HeartRateData()
        val hrIbiList = dataPoint.getValue(ValueKey.HeartRateSet.IBI_LIST)
        val hrIbiStatus = dataPoint.getValue(ValueKey.HeartRateSet.IBI_STATUS_LIST)
        hrData.status = dataPoint.getValue(ValueKey.HeartRateSet.HEART_RATE_STATUS)
        hrData.hr = dataPoint.getValue(ValueKey.HeartRateSet.HEART_RATE)
        if (hrIbiList != null && hrIbiList.size != 0) {
            hrData.ibi = hrIbiList[hrIbiList.size - 1] // Inter-Beat Interval (ms)
        }
        if (hrIbiStatus != null && hrIbiStatus.size != 0) {
            hrData.qIbi = hrIbiStatus.size - 1 // 1: bad, 0: good
        }
        trackerDataNotifier.notifyHeartRateTrackerObservers(hrData)
    }
}
